File: JavaScript/1477-Find Two Non-overlapping Sub-arrays Each With Target Sum.js
/**
 * @param {number[]} arr
 * @param {number} target
 * @return {number}
 */
var minSumOfLengths = function(arr, target) {
// sliding window + DP
    const N = arr.length;
    let dp = new Array(N).fill(Number.POSITIVE_INFINITY);
    let tot = 0;
    let minLen = Number.POSITIVE_INFINITY;
    let start = 0;
    let res = Number.POSITIVE_INFINITY;
    
    for (let i=0; i<N; i++){
        tot += arr[i];
        while (tot > target){
            tot -= arr[start];
            start++;
        }
        if (tot === target){
            let length = i - start + 1;
            if (start > 0 && dp[start-1] !== Number.POSITIVE_INFINITY){
                res = Math.min(res, length + dp[start-1]);
            }
            minLen = Math.min(minLen, length);
        }
        dp[i] = minLen;
    }
    if (res === Number.POSITIVE_INFINITY) return -1;
    else return res;
};

File: C++/1477-Find Two Non-overlapping Sub-arrays Each With Target Sum.cpp
class Solution {
public:
    int minSumOfLengths(vector<int>& arr, int target) {
        int N = arr.size();
        unordered_map<int, int> pos;
        pos[0] = -1;
        int curr = 0, res = N+1, minLength = N+1;
        vector<int> dp(N, -1);  // dp[i] : minimum-length subarray within arr[0:i+1] with target sum

        for (int i=0; i<N; i++) {
            curr += arr[i];
            if (pos.contains(curr-target)) {
                int j = pos[curr-target];
                int currLength = i-j;
                if (j != -1 && dp[j] != -1) {
                    res = min(res, i-j + dp[j]);
                }
                minLength = min(minLength, currLength);
            }

            dp[i] = minLength;
            pos[curr] = i;
        }

        return res==(N+1)?-1:res;
    }
};